<a href="https://colab.research.google.com/github/alexbigfut/Deci/blob/master/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22%D0%94%D0%BE%D0%BC%D0%B0%D1%88%D0%BD%D0%B8%D0%B5_%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D1%8F_%D0%9E%D1%81%D0%BD%D0%BE%D0%B2%D1%8B_Python%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Урок №3. Ввод-вывод и базовые переменные

**Задание №1**

Давай представим, что мы пишем бэкенд для сайта ветеринарной клиники. Нам нужно написать программу, которая будет запрашивать у пользователя вид питомца, его возраст и кличку, а потом выведет все эти данные в одно предложение. Например:

`Это желторотый питон по кличке "Каа". Возраст: 34 года.`

**Решение задания № 1:**

In [ ]:
print ('Какой вид вашего питомца?')
a=input()
print ('какая кличка?')
b=input()
print('какой возраст питомца?')
c=input()
print (('Это ')+ a +(' по кличке ') + b +('. Возраст: ') + c + (' года.'))


**Задание №2**

А теперь мы с тобой напишем форму ввода ответа на тест по биологии для студентов. Он должен запрашивать по порядку этапы развития человека (проверим твое умение гуглить, что тоже очень важно для программиста. ) и в конце вывести все стадии, разделенные знаком `=>`, что будет означать постепенный переход от одного к другому. В следующих уроках мы дополним эту форму до полноценного теста, который будет проверять правильность ответов, а пока - начнем с малого. Напоминаем, что разделить эти данные тебе поможет команда `sep` внутри команды `print`, например, чтобы разделить переменные знаком `+` нужно ввести: 

`print(a1, a2, a3, sep='+')`

Подсказка: последняя стадия развития - Homo sapiens sapiens.

**Решение задания № 2:**

In [ ]:
#дриопитек – рамапитек – австралопитек – человек умелый – человек прямоходящий – неандертальский человек (палеоантроп) – неоантроп (это уже человек современного типа, homo sapiens)
print ('введите через пробел 7 стадий развития человека')
a1,a2,a3,a4,a5,a6,a7=input().split()
print (a1,a2,a3,a4,a5,a6,a7,sep=' => ')

## Урок №4. float, int и арифметические операции

**Задание №1**

Пользователь вводит стороны прямоугольника, выведите его площадь и периметр. На вход программе могут подаваться как целые числа, так и вещественные

**Решение задания № 1:**

In [ ]:
print ('Введите длину сторон прямоугольника a и b в см')
a,b = map(float,input().split())
print ('площадь прямоугольника равна:' , a*b ,'кв.см.')
print ('периметр прямоугольника равен:', (a+b)*2,'см.') 

Введите длину сторон прямоугольника a и b в см
25 25
площадь прямоугольника равна: 625.0 кв.см.
периметр прямоугольника равен: 100.0 см.


**Задание №2**

Дано пятизначное целое число. Напишите алгоритм, который возведёт количество десятков в степень количества единиц. Затем умножит это число на количество сотен. И делит получившееся число на разность количества десятков тысяч и количества тысяч

Например, есть число `46275`

Необходимо возвести 7 (десятки) в степень 5 (единицы), умножить получившееся число на 2 (сотни), и разделить на разность между 4 (десятки тысяч) и 6 (тысячи) то есть (4-6)

В результате необходимо получить вещественное число. В нашем примере это будет: `-16807.0`

**Решение задания № 2:**

In [ ]:
a=input()
b=float(a[0])
c=float(a[1])
d=float(a[2])
e=float(a[3])
f=float(a[4])
g=e**f*d/(b-c)
print (g)

46275
-16807.0


## Урок №5. Логические и условные операторы

**Задание №1**

Пользователь вводит целое число. Выведите его строку-описание вида `"отрицательное четное число"`, `"нулевое число"`, `"положительное нечетное число"`, например, численным описанием числа 190 является строка `"положительное четное число"`. Если число не является четным - выведите сообщение `"число не является четным"`

**Решение задания № 1:**

In [ ]:
print('Введите целое число')
a = int(input())
b= a % 2
if a == 0:
    print('нулевое число')
else:
    if a<0:
        if b!=0:
            print ('отрицательное нечетное число')
        else:
            print('отрицательное четное число')
    else:
        if b!=0:
            print ('положительное нечетное число')
        else:
            print('положительное четное число')

Введите целое число
190
положительное четное число


**Задание №2**

Дано слово из маленьких латинских букв. Сколько там согласных и гласных букв? Гласными называют буквы `«a», «e», «i», «o», «u»`.

Для решения задачи создайте переменную и в неё положите слово с помощью `input()`

А также определите количество каждой из этих гласных букв
Если какой-то из перечисленных букв нет - Выведите False

**Решение задания № 2:**

In [ ]:
x = 'aeiou'
y = input('Введите слово из маленьких латинских букв:\n')

for i in x:
  count = 0
  for r in y:
      if i == r:
        count += 1
  if count > 0:
    print('Букв', i, 'было', count)
  else:
    print('Букв', i, 'False')


Введите слово из маленьких латинских букв:
jdfhvdhfjh
Букв a False
Букв e False
Букв i False
Букв o False
Букв u False


**Задание №3**

Два инвестора - Майкл и Иван хотят вложиться в стартап. Фаундеры сказали, что минимальная сумма инвестиций - `X` долларов, больше инвестировать можно сколько угодно. У Майкла `A` долларов, у Ивана `B` долларов. Если оба могут вложиться - выведите `2`, если только Майкл - `Mike`, если только Иван - `Ivan`, если не могут по отдельности, но вместе им хватает - `1`, если никто - `0`.

**Решение задания № 3:**

In [ ]:
invest = float(input('Введите минимальную сумму инвестиций: '))
cashMike = float(input('введите сумму Майкла: '))
cashIvan = float(input('введите сумму Ивана: '))
if invest <= (cashMike + cashIvan):
  otvet = '1'
  if invest <= cashMike and invest <= cashIvan:
    otvet = '2'
  if invest > cashMike:
      otvet = 'Ivan'
  if invest > cashIvan:
      otvet = 'Mike'  
else:
  otvet = '0' 

print (otvet)

Введите минимальную сумму инвестиций: 100
введите сумму Майкла: 20
введите сумму Ивана: 10
0


## Урок №6. Циклы while и for

**Задание №1**

Cначала вводится число N, затем вводится ровно N целых чисел. Подсчитайте, сколько из них равны нулю, и выведите это количество.

**Решение задания № 1:**

In [ ]:
y = int(input('Введите число: '))
count = 0

while y > 0:
  z = int(input('Введите целое число: '))
  if z == 0:
    count += 1
  y -= 1
print('0 значений было', count)

Введите число: 5
Введите целое число: 23
Введите целое число: 20
Введите целое число: 0
Введите целое число: 32
Введите целое число: 0
0 значений было 2


**Задание №2**

Вводится натуральное число X. Подсчитайте количество натуральных делителей числа X (включая 1 и само число). x ≤ 2e9 (2 миллиарда)

**Решение задания № 2:**

In [ ]:
x = int(input('Введите число: '))

if x <= 2e9:
    if x == 1:
        d = 1
        

    else:
        d = 2
        for i in range(2, int((x/2) + 1)):
            if x % i == 0:
                d += 1

    print(d)
else:
    print('Число больше 2000000000')

Введите число: 2000000003
Число больше 2000000000


**Задание №3**

Вводятся целые числа A и B. Гарантируется, что A ≤ B. Выведите все четные числа на заданном отрезке через пробел.

**Решение задания № 3:**

In [ ]:
a = int(input('Введите a: '))
b = int(input('Введите b должно быть >= a: '))
c = 0
d = ' '

while c <= b:
    if a % 2 != 0:
        a += 1
    for  i in range(a, b+1, 2):
        c = c + a + 2
        d = d + str(i) + ' '
        
    print(d)


Введите a: 13
Введите b должно быть >= a: 113
 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48 50 52 54 56 58 60 62 64 66 68 70 72 74 76 78 80 82 84 86 88 90 92 94 96 98 100 102 104 106 108 110 112 


## Урок №7. Строки

**Задание №1**

На вход подается 1 строка без пробелов. По данной строке определите, является ли она палиндромом (то есть, можно ли прочесть ее наоборот, как, например, слово "шалаш"). Необходимо вывести ”yes”, если строка является палиндромом, и “no” в противном случае.

**Решение задания № 1:**

In [ ]:
a = input('Введите строку:\n')
b = a
l = len(a)-1

for i in a:
    if b[l] != i:
        c = 'No'
        break 
    elif b[l] == i:
        c = 'Yes'
        l -= 1
print(c)
    
    

Введите строку:
asdfgfdsa
Yes


**Задание №2**

Дана строка, длина которой не превосходит 1000. Вам требуется преобразовать все идущие подряд пробелы в один. Выведите измененную строку.

**Решение задания № 2:**

In [ ]:
a = input('Введите строку')
b = ' '
l = len(a)

if l <= 1000:
    a.split()
    b = ' '.join(a.split())
    print(b)
else:
    print('слишком длинная строка ')

Введите строкуsdfsbfj      sdhgjhds    sjdhjshd
sdfsbfj sdhgjhds sjdhjshd


## Урок №8. Списки

**Задание №1**

В первой строке вводится число N. Далее в N строк вводится N чисел (1 ≤ N ≤ 10000), по одному числу на строке. Все числа по модулю не превышают 10e5. Переверните массив чисел. Выведите N чисел - перевернутый массив.

**Решение задания № 1:**

In [ ]:
n = int(input('Введите длину списка: '))
tmp = list(map(int,input().split()))
tmp.reverse()
print(*tmp)


Введите длину списка: 5
1 2 3 4 5
5 4 3 2 1


**Задание №2**

В первую строчку вводится число N (1 ≤ N ≤ 100 000). В следующую строку через пробел вводятся N чисел (1 ≤ Ai ≤ 10e9). Вам требуется написать метод, который получает на вход массив и изменяет его таким образом, чтобы на первом месте стоял последний элемент, на втором - первый, на третьем - второй и т. д. Выведите N чисел - измененный массив. 

**Решение задания № 2:**

In [ ]:
n = int(input('Введите длину списка: '))
tmp = list(map(int,input().split()))
for i in range(len(tmp)):
    if (i < 1) or (i > 10e9):
        k = 0
if  (n > 100000) or (n < 1) or k==0:
    print('не допустимое значение!')
else:
    b = tmp.pop(0)
    tmp.append(b)
    print(*tmp)


Введите длину списка: 3
1 237227682763287367826327863 4
не допустимое значение!


**Задание №3**

На берегу реки стояли n рыбаков, все они хотели перебраться на другой берег. Одна лодка может выдержать не более m килограмм, при этом в лодку помещается не более 2 человек. Определите, какое минимальное число лодок нужно, чтобы перевезти на другой берег всех рыбаков
В первую строку вводится число m (1 ≤ m ≤ 10e6) - максимальная масса, которую может выдержать одна лодка. Во вторую строку вводится число n (1 ≤ n ≤ 100) - количество рыбаков. В следующие N строк вводится по одному числу Ai (1 ≤  Ai ≤ m) - вес каждого путешественника.
Программа должна вывести одно число - минимальное количество лодок, необходимое для переправки всех рыбаков на противоположный берег.

**Решение задания № 3:**

In [ ]:
m = int(input('максимальная масса перевозимая лодкой: '))
n = int(input('введите кол-во рыбаков: '))
boat = 0
weight =[]
if (n >= 1) and (n <= 100):
    print(f"вес: ")
    for i in range(n):
        a = int(input())
        if (a >= 1) and (a <= m):
            weight.append(a)
        else:
            print(f'превышен вес лодки: {m}.')
else:
    print(f'нарушено число рыбаков:  1 ≤ n ≤ 100.')
 
while len(weight):
    boat += 1
    x = weight.pop(0)
    y = m - x
    while i != y:
        y -= 1
        for i in range(len(weight)):
            if i == y:
                weight.pop(i)
#                boat += 1    
print(boat)


максимальная масса перевозимая лодкой: 250
введите кол-во рыбаков: 5
вес: 
120
95
110
100
70
3


## Урок №9. Множества

**Задание №1**

В первую строку вводится число N – количество чисел (1 ≤ N ≤ 100000). Во вторую строку вводится через пробел N чисел, каждое не превышает 2*10e9 по модулю. Требуется выяснить, сколько среди этих чисел различных. Выведите число, равное количеству различных чисел среди данных.

**Решение задания № 1:**

In [ ]:
n = int(input())
if n < 1 or n > 100000:
    print('не допустимое значение')
else:
    tmp = list(map(int,input().split()))
    tmp1 = 
    tmp2 =
    k = 1
    for i in tmp:
        if i<1 or i>2*10e9:
            k = 0

       
if k == 0:
    print('не допустимое значение')
else:
    print(len(tmp))

3
1 2 3
3


**Задание №2**

Вводятся два списка чисел, которые могут содержать до 100000 чисел каждый. Все числа каждого списка находятся на отдельной строке. Выведите, сколько чисел содержится одновременно как в первом списке, так и во втором.

**Решение задания № 2:**

In [ ]:
n = int(input('введите кол-во чисел:'))
sp = []
if n<1 or n>100000:
    print('не допустимое значение')
else:
    for i in range(n):
        uc = input('введите число: ')
        sp.append(uc)
mas = set(sp)

n1 = int(input('введите кол-во чисел:'))
sp1 = []
if n1<1 or n1>100000:
    print('не допустимое значение')
else:
    for i in range(n1):
        uc1 = input('введите число: ')
        sp1.append(uc1)
mas1 = set(sp1)
mas2 = mas.intersection(mas1)
print (len(mas2))

введите кол-во чисел:3
введите число: 1
введите число: 2
введите число: 3
введите кол-во чисел:5
введите число: 2
введите число: 3
введите число: 4
введите число: 5
введите число: 6
2


**Задание №3**

Во входную строку водится последовательность чисел через пробел. Для каждого числа выведите слово ”YES” (в отдельной строке), если это число ранее встречалось в последовательности или ”NO”, если не встречалось.

**Решение задания № 3:**

## Урок №10. Словари

**Задание №1**

Ранее вы выполняли задание связанное с ветеринарной клиникой. В той задаче вам предстояло вывести информацию о питомце на экран. Сейчас вам необходимо создать словарь `pets = {}`

Примерный вид будет следующим:

```
pets = {
  "Имя питомца": {
    'Вид питомца': # придумайте каким образом сюда внести информацию,
    'Возраст питомца': # придумайте каким образом сюда внести информацию,
    'Имя владельца': # придумайте каким образом сюда внести информацию
  }
}
```

У вас должен получиться словарь, с ещё одним словарём внутри. То есть, есть словарь `pets`. Он в себе хранит ещё один словарь, который обозначается именем питомца. Имя питомца также нужно каким-то образом вносить туда.

Задача не будет считаться выполненной, если вы заходите сразу внести информацию, не прибегая в функции `input()`

Например: </br> 
```
pets = {
  "Мухтар": {
    "Вид питомца": "Собака",
    "Возраст питомца": 9,
    "Имя владельца": "Павел" 
  }
}
```
Так должен будет выглядеть результируюший словарь, но первоначальный его вид - пустой. Его необходимо заполнить пользовательским вводом через консоль с помощью функции `input()`, а не вписать значения уже в самом коде.

Возраст питомца должен быть типа `int`
Всё остальное - строки

Так как возраст питомца указывается типом `int`. Необходимо, в соответствии с указанным возрастом выводит `год`, `года` или `лет`. Например:

`Его возраст: 24 года` </br>
`Его возраст: 21 год` </br>
`Его возраст: 19 лет` </br>

И теперь осталось только получить всю информацию о питомце в виде строки, как из задания по `Урок №3. Ввод-вывод и базовые переменные. Задание №1`, но с небольшими изменениями. Для получения информации необходимо воспользоваться методами словаря `keys()` и `values()`:  </br>
`Это желторотый питон по кличке "Каа". Возраст питомца: 19 лет. Имя владельца: Саша`

**Решение задания № 1:**

**Задание №2**

С помощью цикла создайте словарь, в котором ключи будут, например от числа 10, до -5 (включительно). А значениями этих ключей будут сами эти числа возведённые в степени равных этим числам

Например:
```
my_dict = {
  10: 10000000000,
  9: 387420489,
  # и так далее ...
  -5: -0.00032
}
```



**Решение задания № 2:**

## Урок №11. Функции

**Задание №1**

- Создайте функцию, которая принимает в качестве параметра - натуральное целое число.
- Данная функция находит факториал полученного числа

Например, факториал числа `3` это число `6`.

- Теперь создайте список факториалов чисел от получившегося ранее факториала `6`, до `1`.

В итоге, на вход программа получает например число `3`, возвращает число `6` (факториал числа `3`) и вам нужно сделать список из факториалов числа `6` в убывающем порядке. Находим факториал числа `6` - это `720`, затем от числа `5` - это `120` и так далее вплоть до `1`

То есть, результирующий список будет выглядеть в нашем примере так: <br>
`[720, 120, 24, 6, 2, 1]`

**Решение задания № 1:**

**Задание №2**

В `Урок №10. Задание №1` вы создавали словарь с информацией о питомце. Теперь нам нужна "настоящая" база данных для ветеринарной клиники.

Подробный требуемый функционал будет ниже. Пока что справка:

- Создайте функцию `create`
- Создайте функцию `read`
- Создайте функцию `update`
- Создайте функцию `delete`
- Используйте словарь `pets`, который будет предоставлен ниже, либо создайте свой аналогичный

**Функция `create`:**

Данная функция будет создавать новую запись с информацией о питомце и добавлять эту информацию в наш словарь `pets`

**Функция `read`**

Данная функция будет отображать информацию о запрашиваемом питомце в виде: <br>
`Это желторотый питон по кличке "Каа". Возраст питомца: 19 лет. Имя владельца: Саша`

**Функция `update`**

Данная функция будет обновлять информацию об указанном питомце

**Функция `delete`**

Данная функция будет удалять запись о существующем питомце

-- --
Структруа результирующего словаря `pets` будет как и в `Урок №10. Задание №1`, но с небольшим видоизменением: <br>

**Словарь pets**
```python
pets = {
    1:
        {
            "Мухтар": {
                "Вид питомца": "Собака",
                "Возраст питомца": 9,
                "Имя владельца": "Павел"
            },
        },
    2:
        {
            "Каа": {
                "Вид питомца": "желторотый питон",
                "Возраст питомца": 19,
                "Имя владельца": "Саша"
            },
        },
    # и так далее
}

```

Здесь, 1 и 2 - это идентификаторы наших питомцев. Это уникальные ключи, по которым мы сможем обращаться к нашим записям в "базе данных"

Суть будущей программы будет заключаться в следующем:
- Программа будет работать с помощью цикла while с условием `command` != 'stop', то есть до тех пор, пока на предложение ввести команду, пользователь не введёт слово `stop`
- Перед взаимодействием с "базой данных" запрашивается одна из команд в качестве пользовательского ввода. Пусть это будет переменная `command`
- Функция create должна добавлять новую информацию таким образом, чтобы идентификатор увеличивался на единицу. Чтобы у вас была возможность получать последний ключ в словаре воспользуйтесь импортом модуля `collections`. В начале вашей программы пропишите строчку: `import collection`, а в функции `create` в первых строках пропишите следующий код: <br>
```python
def create():
    last = collections.deque(pets, maxlen=1)[0]
```
`last` в данном случае и будет число последнего ключа (или в нашей логике - идентификатора записи). Именно его и необходимо будет увеличивать на единицу при добавлении следующей записи.

Как вам уже известно - суть функций заключается в том, чтобы использовать один и тот же код в нескольких местах. В данной задаче вам предстоит получать информацию о питомце несколько раз. Чтобы не повторяться в коде, вам нужно создать такие функции <br>

`get_pet(ID)`:
```python
def get_pet(ID):
  # функция, с помощью которой вы получите информацию о питомце в виде словаря
  # сделайте проверку, если питомца с таким ID нету в нашей "базе данных"
  # верните в этом случае False
  # а если питомец всё же есть в "базе данных" - верните информацию о нём
  # выглядеть это может примерно так:
  return pets[ID] if ID in pets.keys() else False
```
`get_suffix(age)`:
```python
def get_suffix(age):
  # функция, с помощью которой можно получить суффикс
  # 'год', 'года', 'лет'
  # реализацию этой функции вам предстоит придумать самостоятельно
  # функция будет возвращать соответствующую строку
  return
```
`pets_list()`:
```python
def pets_list():
  # Эта функция будет создана для удобства отображения всего списка питомцев
  # Информацию по каждому питомцу можно вывести с помощью цикла for
```

Обратите внимание, если ID не существует в словаре с питомцами - будет возникать ошибка. Вам можно от неё избавиться, если правильно составить проверочное условие. Здесь не потребуется использовать такие конструкции, как
`try`, `except`, чтобы обработать возникшую ошибку

**Решение задания № 2:**

## Урок №13. Двумерные списки

**Задание №1**

С помощью цикла создайте матрицу вида 10x10 (пример): <br>
```python
[[0, -2, -5, -5, 0, -4, -1, -23, -11, -20], 
[-4, -6, 16, 9, -8, 2, 10, -32, -18, -13], 
[21, 7, -6, -3, 34, -7, -23, 25, 27, 38], 
[7, 29, 38, 19, 55, 48, 46, 49, -8, -46], 
[80, -3, 76, 46, 72, -33, 14, 0, 71, -20], 
[82, 95, 23, -6, 1, 70, 50, 16, 25, 7], 
[52, 36, -25, -35, 18, -36, 93, 11, -44, 9], 
[114, 108, 33, 61, -4, 131, 34, -26, 47, -17], 
[37, 56, 135, -12, 34, 67, -17, 14, 93, -67], 
[-4, 23, 175, -44, 30, -38, 89, 159, 28, 178]]
```

И ещё одну - такой же размерности. Числа в матрице выше привидены в качестве примера (одна из генераций)

Итого у вас должно получиться сперва две матрица одинаковой размерности

И теперь вам нужно сложить эти две матрицы в третью. Формулу сложения матриц, вы можете найти в интернете, либо посетив этот [ресурс](https://studwork.org/spravochnik/matematika/matricy/slojenie-matric)

Итак. Вы должны создать две матрицы

Чтобы заполнить матрицы различными значениями - воспользуйтесь модулем `random`

Для подключения модуля `random` в код используйте строчку, которую нужно поместить в самое начало вашего скрипта
```python
import random
```

**Для справки:** <br>
[Random. Официальная документация по модулю (англ.яз)](https://docs.python.org/3/library/random.html) <br>
[Random. Информация на русском языке](https://pythonworld.ru/moduli/modul-random.html)

Для начала вы можете просто создать переменные `matrix_1` и `matrix_2`

```python
matrix_1 = [[0, -2, -1, -6, -6, 0, -9, -8, -30, -9],
[5, 12, 4, -16, -4, -9, -16, -15, 1, -26],
[13, 39, 14, 23, -4, 40, 32, 6, -8, 23],
[13, -8, 34, 49, 30, 18, 47, 11, -24, 11],
[21, 73, 71, 61, -1, 79, -34, 22, 69, 67],
[75, 25, 25, 39, 100, -12, -21, 81, -10, 87],
[81, 63, 102, 104, 53, -44, 71, -36, -36, -9],
[7, 98, 26, -3, 128, 94, 18, -26, 14, 21],
[65, 128, 80, 124, 27, -32, 73, 59, 19, 34],
[43, 111, 38, 149, 5, 112, 79, 53, 15, 92]]

matrix_2 = [[0, 4, 6, 11, 15, 6, 9, 26, 15, 21],
[-5, 4, -15, -9, -4, 2, -8, 19, -4, -1],
[-2, -39, -19, 14, 22, 5, -34, 15, 16, -9],
[-22, -52, 11, -11, -3, 16, -11, -6, -32, -2],
[-61, -47, -5, -58, 16, -13, 28, -36, -64, 2],
[-29, 23, 19, 2, -14, -87, 7, -88, 39, 7],
[-6, 18, -97, 26, -64, 0, -72, -34, -68, -92],
[-120, -117, -72, -129, -139, 16, -61, 36, -137, -29],
[-112, -83, 7, -119, -132, -129, -143, -154, -23, -34],
[32, -67, -75, -92, 15, -163, 18, 31, -162, -16]]
```

И теперь нужно придумать, как их сложить, чтобы получить `matrix_3`:
```python
matrix_3 = [[0, 2, 5, 5, 9, 6, 0, 18, -15, 12],
[0, 16, -11, -25, -8, -7, -24, 4, -3, -27],
[11, 0, -5, 37, 18, 45, -2, 21, 8, 14],
[-9, -60, 45, 38, 27, 34, 36, 5, -56, 9],
[-40, 26, 66, 3, 15, 66, -6, -14, 5, 69],
[46, 48, 44, 41, 86, -99, -14, -7, 29, 94],
[75, 81, 5, 130, -11, -44, -1, -70, -104, -101],
[-113, -19, -46, -132, -11, 110, -43, 10, -123, -8],
[-47, 45, 87, 5, -105, -161, -70, -95, -4, 0],
[75, 44, -37, 57, 20, -51, 97, 84, -147, 76]]
```

Потренируйтесь сперва в сложении матриц, а затем попробуйте с помощью модуля `random` сгенерировать матрицы подобных размеров и уже на них применить навыки сложения `matrix_1` и `matrix_2`

Задание считается выполненным, если вы напишите алгоритм, который будет уметь как складывать матрицы, так и генерировать матрица различных размерностей. Будь то матрицы `10х10` или `4х3`


**Решение задания № 1:**

## Урок №14. Рекурсия

**Задание №1**


Есть последовательность
```python
my_list = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]
```

Напишите рекурсивную функцию, которая выведет все элементы от первого до последнего и в конце отобразит сообщение
`Конец списка`, если выводить больше нечего. Циклы использовать запрещено

**Решение задания № 1:**

## Урок №15. ООП

**Задание №1**

Есть родительский класс:

In [ ]:
class Transport:

    def __init__(self, name, max_speed, mileage):
        self.name = name
        self.max_speed = max_speed
        self.mileage = mileage

Создайте объект **Autobus**, который унаследует все переменные и методы родительского класса **Transport** и выведете его.

Ожидаемый результат вывода:

**Название автомобиля: Renaul Logan Скорость: 180 Пробег: 12**

**Решение:**


**Задание №2**

Создайте класс **Autobus**, который наследуется от класса **Transport**. 

Дайте аргументу **Autobus.seating_capacity()** значение  по умолчанию **50**.

Используйте переопределение метода.

Используйте следующий код для родительского класса транспортного средства:

In [ ]:
class Transport:
    def __init__(self, name, max_speed, mileage):
        self.name = name
        self.max_speed = max_speed
        self.mileage = mileage

    def seating_capacity(self, capacity):
        return f"Вместимость одного автобуса {self.name}  {capacity} пассажиров"

Ожидаемый результат вывода:

**Вместимость одного автобуса Renaul Logan: 50 пассажиров**



**Решение:**

## Урок №16. Классы и объекты

**Задание №1**

Создайте класс Касса, который хранит текущее количество денег в кассе, у него есть методы:

- top_up(X) - пополнить на X
- count_1000() - выводит сколько целых тысяч осталось в кассе
- take_away(X) - забрать X из кассы, либо выкинуть ошибку, что не достаточно денег

**Задание №2**

Создайте класс Черепашка, который хранит позиции x и y черепашки, а также s - количество клеточек, на которое она перемещается за ход

у этого класс есть методы:

- go_up() - увеличивает y на s
- go_down() - уменьшает y на s
- go_left() - уменьшает x на s
- go_right() - увеличивает y на s
- evolve() - увеличивает s на 1
- degrade() - уменьшает s на 1 или выкидывает ошибку, когда s может стать ≤ 0
- count_moves(x2, y2) - возвращает минимальное количество действий, за которое черепашка сможет добраться до x2 y2 от текущей позиции

## Урок №17. О-нотация

**Задание №1** (ответ дается текстом под кодом)

Какая сложность у этого алгоритма?

In [ ]:
n = int(input())
m = int(input())
cnt = 0
for i in range(n):
	for j in range(m):
		cnt += 1
print(cnt)

3
3
9


**Ответ:**

**Задание №2** (ответ дается текстом под кодом)

Какая сложность у этого алгоритма?

In [ ]:
n = int(input())
ans = n * n + 100
ans -= 200
ans *= ans
if (ans < 1000000):
	ans *= 2

**Ответ:**

**Задание №3** (ответ дается текстом под кодом)

Какая сложность у этого алгоритма?

In [ ]:
n = int(input())
m = int(input())
k = 1000
for i in range(n):
	for j in range(m):
		k += 100

for i in range(m):
	for j in range(m):
		k -= 100

2
3


**Ответ:**

## Урок №18-23. Делаем Игру

Для данного задания нужно использовать программу Visual Studio Code. Вся дальнейшая работа выполняется в ней.


Официальная документация по работе с VSC: https://docs.microsoft.com/ru-ru/learn/modules/python-install-vscode/

Эмодзи можно взять с сайта: https://emojipedia.org/

